lib/um_arclight/package/generator.rb

```diff
@@ -0,0 +1,414 @@
+        elapsed_time = Benchmark.realtime do
+          @fragment = render_fragment(
+            collection: collection,
+            components: components
+          )
+
+          update_navigation_links
+          update_package_html
+        end
+        puts "UM-Arclight generate package : #{collection.id} : build HTML (in #{elapsed_time.round(3)} secs)."
+      end
+
+      def generate_html
+        build_html
+
+        output_filename = generate_output_filename('.html')
+        FileUtils.makedirs(File.dirname(output_filename)) unless Dir.exist?(File.dirname(output_filename))
+
+        File.open(output_filename, 'w') do |f|
+          f.puts doc.serialize
+        end
+      end
+
+      def build_pdf
+        # build the source in tmp
+        FileUtils.mkdir_p(working_path_name)
+        Dir.chdir(working_path_name)
+        FileUtils.mkdir_p('assets')
+
+        elapsed_time = Benchmark.realtime do
+          update_package_html_pdf
+          update_package_styles_pdf
+          update_package_scripts_pdf
+          # set the media
+          doc.root['data-media'] = 'print'
+        end
+        puts "UM-Arclight generate package: #{collection.id} : update HTML for PDF (in #{elapsed_time.round(3)} secs)."
+      end
+
+      def generate_pdf # rubocop:disable Metrics/MethodLength
+        generate_html if @doc.nil?
+
+        build_pdf
+
+        local_html_filename = "#{collection.id}.local.html"
+        File.open(local_html_filename, 'w') do |f|
+          f.puts doc.serialize
+        end
+
+        output_filename = generate_output_filename('.pdf')
+        FileUtils.mkdir_p(File.dirname(output_filename))
+
+        elapsed_time = Benchmark.realtime do
+          Puppeteer.launch(headless: true, args: ['--no-sandbox', '--disable-setuid-sandbox']) do |browser|
+            page = browser.new_page
+            page.goto("file:#{working_path_name}/#{identifier}.local.html", wait_until: 'networkidle2')
+            page.pdf(
+              path: output_filename,
+              print_background: true,
+              omit_background: false,
+              display_header_footer: false,
+              timeout: (2 * 300_000),
+              footer_template: '<div style="font-weight: bold">Generated by findingaids.lib.umich.edu</div>',
+              margin: {
+                top: 50,
+                right: 100,
+                bottom: 70,
+                left: 70
+              }
+            )
+          end
+        end
+
+        File.unlink(local_html_filename) unless ENV.fetch('DEBUG_GENERATOR', 'FALSE') == 'TRUE'
+
+        puts "UM-Arclight generate package: #{collection.id} : puppeteer render (in #{elapsed_time.round(3)} secs)."
+      end
+
+      private
+
+      def generate_output_filename(ext)
+        filename = "#{DulArclight.finding_aid_data}/pdf/#{collection.repository_id}/#{collection.id}#{ext}"
+        filename = File.join(Rails.root, filename) if filename.start_with?('./')
+        filename
```

Comment:
This is another opportunity for using Pathname, which I've found both convenient and less prone to weird space and slash errors. Not needed now; an idea for later improvement.